static/app/utils/profiling/flamegraph.ts

```diff
@@ -351,4 +357,70 @@ export class Flamegraph {
 
     return matches;
   }
+
+  collapseSystemFrames() {
+    const {root, frames, depth} = buildCollapsedFlamegraph(collapse(this.root));
+    this.root = root;
+    this.depth = depth;
+    this.frames = frames;
+  }
+}
+
+const isSystemFrame = (frame: FlamegraphFrame) => !frame.node.frame.is_application;
+
+function collapse(frame: FlamegraphFrame) {
+  frame.children = frame.children.reduce((acc, child) => {
+    acc.push(collapse(child));
+    return acc;
+  }, [] as FlamegraphFrame[]);
+
+  if (!frame.parent) {
+    return frame;
+  }
+
+  const children = frame.children;
+
+  if (children.length === 1) {
+    const [child] = children;
+    if (isSystemFrame(frame) && isSystemFrame(child)) {
+      if (!child.collapsed) {
+        child.collapsed = 1;
+      } else {
+        child.collapsed++;
+      }
+      child.parent = frame.parent;
+      frame = child;
+    }
```

Comment:
This is collapsing a chain of system frames into the deepest frame right? i.e. `a -> b -> c -> d` if `a` is application and `b`, `c`, `d` are system, them it turns into `a -> d`.

Reply:
yes, that's correct

Reply:
we could retain all of this context if we wanted

instead of collecting the count of `collapsed`, we could instead keep a set of frames that were collapsed and render something that shows the user what was collapsed.

for example, if we change the text rendering to be `n system frames` the hover over tool tip could show you the list of frames which were collapsed

we could even show a simplified notation as you've used above; `(+n) -> b()` or `a() -> +n -> b()`

Reply:
i guess my question is if we want that to become

`a -> b (+ n system frames)`

or

`a -> n system frames -> d`

or

something else. I had initially imagined the first one but that's up for discussion

Reply:
we can play around with it, i'll change `collapsed` to actually collect the frames for now